runtime/near-vm-runner/src/cache.rs

```diff
@@ -71,19 +71,26 @@ fn compile_and_serialize_wasmer(
     Ok(module)
 }
 
-fn deserialize_wasmer(serialized: &[u8]) -> Result<wasmer_runtime::Module, VMError> {
+/// Deserializes contract or error from the binary data. Signature means that we could either
+/// return module or cached error, which both considered Ok(), or encounter an error during
+/// deserialization process.
+fn deserialize_wasmer(
+    serialized: &[u8],
+) -> Result<Result<wasmer_runtime::Module, VMError>, VMError> {
```

Comment:
Can we have a dedicated error enum with: DeserializationError, CompilationError(VMError) instead of Result<Result<_, VMError>, VMError> (it is really hard to reason about this return value type)

Reply:
That's hard question. If we think about our API as "stack", than compilation error remembered earlier and stored in the database is indeed success, and not an error in deserialization logic, while it's an error in higher level function. SO IMO current signature is pretty representative to that notion, and fortunately it doesn't escape cache.rs anyway.